cpp/src/link_prediction/similarity_impl.cuh

```diff
@@ -81,84 +81,148 @@ rmm::device_uvector<weight_t> similarity(
     rmm::device_uvector<weight_t> weighted_out_degrees =
       compute_out_weight_sums(handle, graph_view, *edge_weight_view);
 
-    per_v_pair_transform_dst_nbr_intersection(
-      handle,
-      graph_view,
-      *edge_weight_view,
-      vertex_pairs_begin,
-      vertex_pairs_begin + num_vertex_pairs,
-      weighted_out_degrees.begin(),
-      [functor] __device__(auto a,
-                           auto b,
-                           auto weight_a,
-                           auto weight_b,
-                           auto intersection,
-                           auto intersected_properties_a,
-                           auto intersected_properties_b) {
-        weight_t sum_of_min_weight_a_intersect_b = weight_t{0};
-        weight_t sum_of_max_weight_a_intersect_b = weight_t{0};
-        weight_t sum_of_intersected_a            = weight_t{0};
-        weight_t sum_of_intersected_b            = weight_t{0};
-
-        auto pair_first = thrust::make_zip_iterator(intersected_properties_a.data(),
-                                                    intersected_properties_b.data());
-        thrust::tie(sum_of_min_weight_a_intersect_b,
-                    sum_of_max_weight_a_intersect_b,
-                    sum_of_intersected_a,
-                    sum_of_intersected_b) =
-          thrust::transform_reduce(
+    if constexpr (functor_t::is_order_two) {
```

Comment:
What is order_two here?

if is_order_two == true, we do this
```
return thrust::make_tuple(prop_a * prop_a, prop_b * prop_b, prop_a * prop_b);
```

and if is_order_two == false, we do the following?

```
return thrust::make_tuple(min(prop_a, prop_b), max(prop_a, prop_b), prop_a, prop_b);
```

Reply:
cosine similarity requires second order terms. I couldn't come up with a better name.

Reply:
Naming is fine, but we should better avoid code replication if the difference is only in `prop_a * prop_a, prop_b * prop_b` vs `min(prop_a, prop_b), max(prop_a, prop_b)`. Code replication makes it tougher to figure out what order_two means.

Reply:
Changed.